protos/sigstore_verification.proto

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+package dev.sigstore.verification.v1;
+
+import "sigstore_common.proto";
+import "sigstore_trustroot.proto";
+import "sigstore_bundle.proto";
+
+option go_package = "github.com/sigstore/proto/verification/v1";
+option java_package = "dev.sigstore.proto.verification.v1";
+option java_multiple_files = true;
+option java_outer_classname = "VerificationProto";
+
+// The identity of a signer.
+message Identity {
```

Comment:
I may be wrong here, but regardless if it's an email or a "GitHub workflow name", they all go into the same field (X509v3 SAN), but with a "type" prefix, like `email:` or `URI`. So the intent was to always match against the SAN.

Reply:
https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md

Reply:
But that does not specify what goes into the SAN, only the various extensions Fulcio have. 
Let me dig more into where the certificate is create and the SAN is populated, I think it may be here: https://github.com/sigstore/fulcio/blob/67c6e2f8a943d0d26337716528fbd7f8eb521908/pkg/server/grpc_server.go#L82
But I have to jump on a meeting now, I'll follow this up more in detail tomorrow.

Reply:
https://github.com/sigstore/fulcio/blob/67c6e2f8a943d0d26337716528fbd7f8eb521908/pkg/identity/github/principal.go#L106
https://github.com/sigstore/fulcio/blob/67c6e2f8a943d0d26337716528fbd7f8eb521908/pkg/identity/email/principal.go#L70

Examples where the workflow name and the email address are added to the SAN. But now realizing the SAN is typed we should add that into the `Identity` message. Makes sense?

Reply:
I also have been opposed to using SAN anywhere, because it's a detail of x509. I'd rather use `identity` for a message name, and let the client handle the logic of extracting the subject from the relevant location. Lots of related conversation in https://github.com/sigstore/cosign/issues/1964 about not making this typed

We should consider GitHub Actions identities too, but I'm fine punting that til later, because we will be working on creating a standard set of claims for CI soon.

Reply:
> So the intent was to always match against the SAN.

IMO that's not the right abstraction to expose. I've been thinking of "identity" as basically a predicate over the certificate. *Most* of what we care about is the SAN/issuer, but you could imagine caring about those other Fulcio-specific extensions.

To be clear I'm okay with this as-is, just thinking long term

Reply:
I've updated this now to be more flexible going forward (`oneof` to capture the type of identity we will verify, only SAN is added now). Resolving this.

Reply:
Sorry to be a little annoying...I actually don't think `oneof` is appropriate here.

I may want to check the SAN *and* one of these other fields.

I'd be totally fine with just a comment noting that we may in the future add ways to check other conditions on the cert

Reply:
Agreed that it's not a oneof. Do we want to extract every identity-based field (SAN, and other custom OIDs) into a named field?

Reply:
Actually, it makes sense. Would we consider something like this:
```protobuf
message Identity {
        // The X.509v3 issuer extension (OID 1.3.6.1.4.1.57264.1.1)
        string issuer = 1;
        SubjectAlternativeName san = 2;
        optional CustomOID1 = 2; // e.g. 1.3.6.1.4.1.57264.1.2 (GitHub workflow trigger)
        optional CustomOID2 = 3; e.g. 1.3.6.1.4.1.57264.1.6 (GitHub worfkflow ref)
}
```

So we require issuer and SAN, but have the a set of custom OIDs (like all the ones from Fulcio, maybe some others too) as optional. The client will specify which to use, then the verifier will loop over all the materialized OIDs and verify? I think this is what @haydentherapper proposed in the last comment?

Reply:
That looks good to me.

Reply:
This is updated now, instead of a map, I added a list of `{oid, bytes}` instead. The `oid` is now a typed message of `[]int`. Resolving this.